Converts holodeck to use dmmsuite rather than copying areas

mikomyazaki opened the pull request:
:cl: mikomyazaki
bugfix: Holodeck programs will now spawn mobs correctly, go fight a holocarp!
bugfix: Thunderdome ready buttons work properly.
maptweak: Can now only change the holodeck program once per minute (lag prevention measure)
rscadd: The holodeck can now simulate a firing range, with lifelike targets!
/:cl:

This is nearly complete, so I'm putting it up to get some feedback on the way I've gone about it.

Intention is to adjust the holodeck to load map templates instead of copying an area, since copying an area causes various problems with a lot of atoms not acting how they should. Holodeck maps now live in their own .dmms under maps/torch/game/holodeck with their own datums that should help to make more interesting holodeck maps in the future.

This depends upon #26644 

Fixes #25142
Fixes #13714 
Fixes #23040
Fixes #26714 
Fixes #27877
Fixes #28105

To Do:
- [x] Make non-holo stuff inside holo-containers not get qdel'd when the holo-container leaves the holodeck.
- [x] Make the gravity toggle button work
- [x] Make sure holo-objs can't be damaged, and either can't dismantle or gives zero stuff
- [x] Make sure holo-things can't leave the holo-place

PsiOmegaDelta commented on code/game/atoms.dm:
```diff
@@ -13,6 +13,7 @@
 	var/datum/reagents/reagents // chemical contents.
 	var/list/climbers
 	var/climb_speed_mult = 1
+	var/holographic = FALSE
```
Would be greatly preferable if this was an extension which properly handles setting alpha, registering movement checks to ensure they don't leave the current area, etc.

mikomyazaki replied:
Extension'd it.

mikomyazaki commented:
Dealt with the problem of containers not working properly. Ready for review, I think.

PsiOmegaDelta commented on code/modules/maps/map_template.dm:
```diff
@@ -148,7 +149,10 @@
+	if(M && get_metadata)
+		return M
+	else
+		return TRUE
```
Why just not always return this "metadata"?

mikomyazaki replied:
Now always returns the metadata.

PsiOmegaDelta commented on code/modules/maps/map_template.dm:
```diff
@@ -132,10 +132,11 @@
 	var/initialized_areas_by_type = list()
 	for (var/mappath in mappaths)
-		var/datum/map_load_metadata/M = maploader.load_map(file(mappath), T.x, T.y, T.z, cropMap=TRUE, clear_contents=(template_flags & TEMPLATE_FLAG_CLEAR_CONTENTS), initialized_areas_by_type = initialized_areas_by_type)
+		M = maploader.load_map(file(mappath), T.x, T.y, T.z, cropMap=TRUE, clear_contents=(template_flags & TEMPLATE_FLAG_CLEAR_CONTENTS), initialized_areas_by_type = initialized_areas_by_type)
```
So if multiple map files are loaded only the last generated metadata output is returned?

mikomyazaki replied:
Will now store all the metadata from each loaded map in a list and return that.
/datum/map_load_metadata now keeps the map filename so list entries can be differentiated between if necessary.

PsiOmegaDelta commented on baystation12.dme:
```diff
@@ -3106,12 +3107,33 @@
+#include "maps\torch\game\holodeck\theatre\theatre.dm"
+#include "maps\torch\game\holodeck\thunderdomecourt\thunderdomecourt.dm"
+#include "maps\torch\game\holodeck\volleyball\volleyball.dm"
+#include "maps\torch\game\holodeck\wildlife\wildlife.dm"
```
Why are holodecks designated to be Torch-specific included in the general DME?

mikomyazaki replied:
Removed from the dme.

PsiOmegaDelta commented on maps/torch/game/holodeck/holodeck_programs.dm:
```diff
@@ -0,0 +1,15 @@
+	if (suffixes)
+		mappaths = list()
+		for (var/suffix in suffixes)
+			mappaths += (prefix + suffix)
```
Not calling `..()`
The suffixes-setup looks like a copy-paste job from `/datum/map_template/ruin`

mikomyazaki replied:
Fixed. 
Doesn't need to support multiple mappaths.
One map per holodeck program.

No actual need for the whole suffix thing. Just directly defined a mappaths list in the datums now.

mikomyazaki replied:
Also made suffix not a list in all the program specific datums.

mikomyazaki commented:
Going to have to wait until #26666 is in to continue this.

PsiOmegaDelta commented on code/modules/holodeck/HolodeckControl.dm:
```diff
@@ -23,10 +25,17 @@
 	var/programs_list_id = null
 	var/list/supported_programs = list()
 	var/list/restricted_programs = list()
+	var/datum/map_load_metadata/M = null
```
Keeping a reference to the loaded metadata will in turn prevent referenced instances from GCing cleanly unless handling the deleted event

PsiOmegaDelta replied:
M itself should also be handled in `/Destroy()`

mikomyazaki replied:
I think I've resolved this? Unsure if I did it right.

comma commented on code/datums/extensions/holographic.dm:
```diff
@@ -0,0 +1,55 @@
+	. = ..()
+
+/datum/extension/holographic/proc/checkHolographicContents(var/atom/item) //recursively check for holographic items within containers and delete them, or non-holographic items within holographic containers and drop them.
+	if(item?.contents)
```
fyi you don't really need to check list before looping, for var in null will just not loop and continue safely.

mikomyazaki commented:
Now works with Psi's new map template loading stuff. Although has some exciting new bugs for me to squash, like taking 20s to load a map.

mikomyazaki commented:
Okay, bugs dealt with (I think) ready for review.

PsiOmegaDelta commented:
https://github.com/Baystation12/Baystation12/blob/dev/code/modules/holodeck/HolodeckObjects.dm has multiple cases of  `New(`)` overrides which do not call `..()`, causing the atom init test to fail

PsiOmegaDelta commented on code/modules/holodeck/HolodeckControl.dm:
```diff
@@ -154,21 +163,27 @@
 	emergencyShutdown()
+	for(var/atom/item in M.atoms_to_initialise)
+		qdel(item)
+	M = null
```
Better would be to `QDEL_NULL(M)` and have `/datum/map_load_metadata` clear its own lists in `/Destroy()`
`/datum/map_load_metadata` probably shouldn't delete the contents of them

PsiOmegaDelta replied:
Belay that

mikomyazaki replied:
Resolving this because you suggest a better fix later on, which I will implement.

PsiOmegaDelta commented on code/modules/holodeck/HolodeckControl.dm:
```diff
@@ -265,52 +257,37 @@
+			if(istype(holo,/atom/movable))
+				set_extension(holo,/datum/extension/holographic,/datum/extension/holographic/movable)
+			else
+				set_extension(holo,/datum/extension/holographic,/datum/extension/holographic)
```
Better is probably for `/datum/extension/holographic` to handle the difference between moving and un-moving atoms

mikomyazaki replied:
Done...  but please check this incase I'm misunderstanding how overwriting an a parent extension with set_extension works.

PsiOmegaDelta commented on code/modules/holodeck/HolodeckControl.dm:
```diff
@@ -265,52 +257,34 @@
-	for(var/obj/holo_obj in holographic_objs)
-		holo_obj.alpha *= 0.8 //give holodeck objs a slight transparency
-		holo_obj.holographic = TRUE
+	M = HP.load(linkedholodeck_corner)[1] //should only ever spawn one map at once with the holodeck
```
Actually, having looked at it further the better option might be to not store the `/datum/map_load_metadata` at all
Storing it long-term means that one either has to register to the destruction events or keep weakrefs

Registering to events like that would prevent automatic garbage collection of the `/datum/map_load_metadata` instance due to reference counting and would force anyone loading holodecks to remember to qdel it.

Weakrefs is a possible solution but might require a fair amount of `resolv()`ing, which might be somewhat performance intensive.

Instead, just copy the `atoms_to_initialise` list, loop over it to set everything up, and then also register to the `destroyed_event` to remove deleted objects as necessary.

mikomyazaki replied:
Done, I think.

PsiOmegaDelta commented on maps/torch/game/holodeck/holodeck_programs.dm:
```diff
@@ -0,0 +1,9 @@
+	var/list/ambience
+
+/datum/map_template/holodeck/New()
+	..()
```
A somewhat superfluous override now that it's empty

mikomyazaki replied:
Done.

PsiOmegaDelta commented on code/modules/holodeck/HolodeckControl.dm:
```diff
@@ -109,8 +118,8 @@
-			if(prog in GLOB.using_map.holodeck_programs)
-				loadProgram(GLOB.using_map.holodeck_programs[prog])
+			if(prog in supported_programs)
+				loadProgram(SSmapping.holodeck_templates[prog])
```
Curious as to why you're not using `supported_programs[prog]` when you go through the effort to do the assignment on line 40

mikomyazaki replied:
No reason. Just missed renaming it after I did the supported_programs thing. Changed it.

PsiOmegaDelta commented on code/modules/holodeck/HolodeckControl.dm:
```diff
@@ -265,52 +257,34 @@
+			set_extension(holo,/datum/extension/holographic,/datum/extension/holographic)
+		for(var/obj/holo in M.atoms_to_initialise)
+			holographic_objs |= holo
+		for(var/mob/holo in M.atoms_to_initialise)
```
Seems a little unnecessary to loop over the same list 3 times when you can just type-check inside the first loop (as that's basically what these 2 additional loops do)

mikomyazaki replied:
Changed this as suggested.

PsiOmegaDelta commented on code/datums/extensions/holographic.dm:
```diff
@@ -0,0 +1,57 @@
+/datum/extension/holographic/New(var/atom/holder)
+	..()
+	if(istype(src,/atom/movable) && !has_extension(src,/datum/extension/holographic/movable))
+		set_extension(src,/datum/extension/holographic,/datum/extension/holographic/movable)
```
Just register to the movement event if it's an `/atom/movable` instead of setting another extension
Also remember to unregister in `/Destroy()`

mikomyazaki replied:
Done.

PsiOmegaDelta commented on code/datums/extensions/holographic.dm:
```diff
@@ -0,0 +1,57 @@
+	makeContentsHolographic(holder)
+
+/datum/extension/holographic/proc/makeContentsHolographic(var/atom/item)
+	for(var/atom/held_item in item.contents && !istype(held_item,/obj/effect/landmark))
```
This doesn't look like a loop that'll work, and even if it does
```
for(var/object in item)
    var/atom/A = object
    if(!A.simulated)
        continue
    set_extension(A, /datum/extension/holographic, /datum/extension/holographic)
```

mikomyazaki replied:
Done.

PsiOmegaDelta commented on code/datums/extensions/holographic.dm:
```diff
@@ -0,0 +1,57 @@
+/datum/extension/holographic/movable/proc/OnMovement(var/atom/mover, var/old_location, var/new_location)
+	if(get_area(old_location) != get_area(new_location))
+		if(mover?.contents)
+			checkHolographicContents(mover)
```
Seems like a better name would be `/DestroyHolographicContents()`

PsiOmegaDelta replied:
No need to waste time checking `contents` here either. The for-loop in the proc handles that fine

mikomyazaki replied:
Sure. Changed. (both bits)

PsiOmegaDelta commented on code/datums/extensions/holographic.dm:
```diff
@@ -0,0 +1,50 @@
+/datum/extension/holographic/proc/OnMovement(var/atom/mover, var/old_location, var/new_location)
+	if(get_area(old_location) != get_area(new_location))
+		DestroyHolographicContents(mover)
+		qdel(mover)
```
``` 
/datum/extension/holographic/New(var/atom/holder)
	..()
    holder.MakeHolographic(src)

/datum/extension/holographic/Destroy(var/atom/holder)
    holder.UnmakeHolographic(src)
	. = ..()

/atom/proc/MakeHolographic(/datum/extension/holographic/H)
    alpha *= H.holo_alpha
    for(var/object in H)
        set_extension(object, /datum/extension/holographic, H.type)

/atom/movable/MakeHolographic(/datum/extension/holographic/H)
    ..()
    GLOB.moved_event.register(src, H, /datum/extension/holographic/movable/proc/OnMovement)

/mob/MakeHolographic(/datum/extension/holographic/H)
    ..()
    GLOB.death_event.register(src, H, /datum/extension/holographic/movable/proc/DestroyHologram)

// Also implement the reverse of these procs as /proc/UnmakeHolographic

 /datum/extension/holographic/movable/proc/OnMovement(var/mover, var/old_loc, var/new_loc)
    if(get_area(old_location) != get_area(new_location))
        DestroyHologram()

/datum/extension/holographic/movable/proc/DestroyHologram(var/atom/movable/hologram)
    qdel(hologram) // We assume that contained objects deal with themselves just fine, due to having their own hologram extension handling
```

Just tossed together from the top of my head but might be a little cleaner

comma commented:
Also 
```
code/controllers/subsystems/mapping.dm:73:error: /datum/map_template/holodeck: undefined type path
code/modules/holodeck/HolodeckControl.dm:242:error: HP: undefined type: HP
code/modules/holodeck/HolodeckControl.dm:245:error: HP: undefined type: HP
code/modules/holodeck/HolodeckControl.dm:245:error: HP.name: undefined type: HP.name
code/modules/holodeck/HolodeckControl.dm:272:error: HP.load: undefined type: HP.load
code/modules/holodeck/HolodeckControl.dm:282:error: HP.ambience: undefined type: HP.ambience
code/modules/holodeck/HolodeckControl.dm:283:error: HP.ambience: undefined type: HP.ambience
```

comma commented:
Legit Travis fail
```
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/red) ([0x203f412]) (the floor) (104,109,40) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/green) ([0x203f41c]) (the floor) (104,100,40) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/red) ([0x203f422]) (the floor) (105,109,40) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/green) ([0x203f42c]) (the floor) (105,100,40) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/red) ([0x203f432]) (the floor) (106,109,40) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/green) ([0x203f43c]) (the floor) (106,100,40) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/red) ([0x203f442]) (the floor) (107,109,40) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/green) ([0x203f44c]) (the floor) (107,100,40) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/red) ([0x203f452]) (the floor) (108,109,40) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/green) ([0x203f45c]) (the floor) (108,100,40) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/red) ([0x203f72b]) (the floor) (104,109,49) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/green) ([0x203f735]) (the floor) (104,100,49) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/red) ([0x203f73b]) (the floor) (105,109,49) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/green) ([0x203f745]) (the floor) (105,100,49) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/red) ([0x203f74b]) (the floor) (106,109,49) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/green) ([0x203f755]) (the floor) (106,100,49) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/red) ([0x203f75b]) (the floor) (107,109,49) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/green) ([0x203f765]) (the floor) (107,100,49) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/red) ([0x203f76b]) (the floor) (108,109,49) (/turf/simulated/floor/holofloor/tiled)
## UNIT_TEST ##: [SUBSYSTEM - ATOMS: All atoms shall be initialized.]: Uninitialized atom: The holosword (/obj/item/weapon/holo/esword/green) ([0x203f775]) (the floor) (108,100,49) (/turf/simulated/floor/holofloor/tiled)
```

mikomyazaki commented:
Going to be away from things for about a week. Will reopen when I'm back.

mikomyazaki commented:
Nothing new yet, just rebasing & such.

PsiOmegaDelta commented on code/datums/extensions/holographic.dm:
```diff
@@ -0,0 +1,50 @@
+	holder.alpha *= holo_alpha
+	makeContentsHolographic(holder)
+
+/datum/extension/holographic/proc/makeContentsHolographic(var/atom/item)
```
Inconsistent casing with other procs in the file

mikomyazaki replied:
Fixed.

PsiOmegaDelta commented on code/datums/extensions/holographic.dm:
```diff
@@ -0,0 +1,50 @@
+
+/datum/extension/holographic/proc/DestroyHolographicContents(var/atom/item) //recursively check for holographic items within containers and delete them, or non-holographic items within holographic containers and drop them.
+	if(item)
+		for(var/obj/held_item in item.contents)
```
```
for(var/hi in item)
    var/atom/movable/held_item = hi
```

mikomyazaki replied:
Fixed.

PsiOmegaDelta commented on code/datums/extensions/holographic.dm:
```diff
@@ -0,0 +1,50 @@
+					DestroyHolographicContents(held_item)
+				qdel(held_item)
+			else
+				held_item.dropInto(held_item.locs[1])
```
`held_item.dropInto(holder.loc)`

mikomyazaki replied:
Fixed.

PsiOmegaDelta commented on code/datums/extensions/holographic.dm:
```diff
@@ -0,0 +1,50 @@
+				held_item.dropInto(held_item.locs[1])
+
+/datum/extension/holographic/proc/ValidTarget(var/atom/A)
+	..()
```
There's no purpose to calling `..()` is a base proc

PsiOmegaDelta commented on code/datums/extensions/holographic.dm:
```diff
@@ -0,0 +1,50 @@
+			else
+				held_item.dropInto(held_item.locs[1])
+
+/datum/extension/holographic/proc/ValidTarget(var/atom/A)
```
This proc never appears to be called

mikomyazaki replied:
I mean for this to override the `ValidTarget` proc from `/mob/living/simple_animal/hostile/proc/ValidTarget` so that they can't target anything outside the holodeck. Is this not the right way to go about that?

MistakeNot4892 replied:
For reference (necropost because I'm reviewing this prior to stealing it for Nebula), proc overrides need to override the proc on the _type_. In this case you're defining a new proc on /datum/extension/holographic called `ValidTarget()`.

PsiOmegaDelta commented on code/datums/extensions/holographic.dm:
```diff
@@ -0,0 +1,50 @@
+	makeContentsHolographic(holder)
+
+/datum/extension/holographic/proc/makeContentsHolographic(var/atom/item)
+	for(var/object in item.contents)
```
`for(var/object in item)´

mikomyazaki replied:
Fixed.

mikomyazaki commented:
Fixed some review stuff, and made it actually work again. More or less complete and functional now, excepting a couple of review points to handle.

mikomyazaki commented:
Yay. Travis passing.

mikomyazaki commented:
Updated for psi's extension setting simplification PR.

mikomyazaki commented:
Now additionally includes a holographic firing range map to allow testing of new features relating to mobs etc.

afterthought2 commented:
@PsiOmegaDelta, this seems ready for review. Can you take another look?

mikomyazaki commented:
Merge conflicts dealt with. Travis *might* throw something at me.

This is ready to review / maybe to ship.

mikomyazaki committed:
Converts holographicness to an extension.
Adds holographic firing range.
Makes holodeck use dmmsuite
Holodeck control has a nanoui interface now.

mikomyazaki commented:
Added a nano-ui interface for the control computer.

Spookerton commented:
author can reopen this or someone else can pick it up.
